### PR TITLE
Move connect decision into protocol layer

### PR DESCRIPTION
**Pull request by pinges**
The aim of this PR is to move the control over which peers Besu is connecting to from the RLPX layer to the protocol layer. Currently the RLPX layer is making sure that we do not have too many connections. Once the maximum number of connections has been reached new connections are not created (unless randomPeerPriority is configured). On the RLPX layer we do not have information like what protocols are used at the moment or whether we are syncing and how we are syncing. 
This PR does not change the selection of connections, it leaves the logic the same, but it allows future changes to the selection process by moving the decision into the protocol layer, where the requirements for connections can be based on more information. 
The following classes have major changes:

- RlpxAgent: removed code that limited the number of connections, now refers to the protocol layer 
- EthPeers: Now contains the logic removed from RlpxAgent to limit connections
- EthPeer, EthProtocolManager, SnapProtocolManager, BftProtocolManager

Signed-off-by: Stefan <stefan.pingel@consensys.net>

### CODE REVIEW

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java [hunk 1]
```diff
@@ -66,56 +65,41 @@ public class RlpxAgent {
...
 
   private final PeerRlpxPermissions peerPermissions;
   private final PeerPrivileges peerPrivileges;
-  private final int upperBoundConnections;
-  private final int lowerBoundConnections;
-  private final boolean randomPeerPriority;
-  private final int maxRemotelyInitiatedConnections;
-  // xor'ing with this mask will allow us to randomly let new peers connect
-  // without allowing the counterparty to play nodeId farming games
-  private final Bytes nodeIdMask = Bytes.random(SECPPublicKey.BYTE_LENGTH);
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java:78]: did this logic move elsewhere or don't we need it any more

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java [hunk 2]
```diff
@@ -66,56 +65,41 @@ public class RlpxAgent {
...
-    this.maxRemotelyInitiatedConnections =
-        Math.min(upperBoundConnections, maxRemotelyInitiatedConnections);
-
-    // Setup metrics
-    connectedPeersCounter =
-        metricsSystem.createCounter(
-            BesuMetricCategory.PEERS, "connected_total", "Total number of peers connected");
-
-    metricsSystem.createIntegerGauge(
-        BesuMetricCategory.ETHEREUM,
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java:115]: these metrics moved to EthPeers

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 3]
```diff
@@ -58,40 +67,54 @@ public class EthPeers {
...
     this.clock = clock;
     this.permissioningProviders = permissioningProviders;
-    this.maxPeers = maxPeers;
     this.maxMessageSize = maxMessageSize;
     this.bestPeerComparator = HEAVIEST_CHAIN;
+    this.localNodeId = localNodeId;
+    this.peerLowerBound = peerLowerBound;
+    this.peerUpperBound = peerUpperBound;
+    this.maxRemotelyInitiatedConnections = maxRemotelyInitiatedConnections;
+    this.randomPeerPriority = randomPeerPriority;
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:117]: what's the difference between the peer_count metric and the connected_total?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java [hunk 4]
```diff
@@ -157,50 +141,58 @@ public CompletableFuture<Void> stop() {
...
+      throw new RuntimeException(e);
+    }
   }
 
   public int getConnectionCount() {
-    return Math.toIntExact(streamConnections().count());
+    try {
+      return (int) getAllActiveConnectionsCallback.call().count();
+    } catch (final Exception e) {
+      throw new RuntimeException(e);
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java:164]: when would we get this exception?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java [hunk 5]
```diff
@@ -225,106 +217,78 @@ public CompletableFuture<PeerConnection> connect(final Peer peer) {
...
-            .map(
-                updatedPeers ->
-                    updatedPeers.stream()
-                        .map(peer -> connectionsById.get(peer.getId()))
-                        .filter(connection -> !isNull(connection))
-                        .collect(Collectors.toList()))
-            .orElse(new ArrayList<>(connectionsById.values()));
+    final Stream<PeerConnection> connectionsToCheck;
+    if (peers.isPresent()) {
+      final List<Bytes> changedPeersIds =
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java:277]:
```suggestion
      final List<Bytes> matchingPeerIds =
```

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java [hunk 6]
```diff
@@ -348,36 +312,29 @@ private CompletableFuture<PeerConnection> initiateOutboundConnection(final Peer
...
+    this.getAllConnectionsCallback = callback;
+  }
+
+  public void setGetAllActiveConnectionsCallback(final Callable<Stream<PeerConnection>> callback) {
+    this.getAllActiveConnectionsCallback = callback;
+  }
+
   private void handleIncomingConnection(final PeerConnection peerConnection) {
+    // TODO: when we get here, we are already connected and have spent the time and effort to go
+    // through the handshake (expensive). We might want to put in a check before doing the handshake
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java:329]: is there a follow up task for this TODO

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java [hunk 7]
```diff
@@ -386,171 +343,12 @@ && getConnectionCount() >= upperBoundConnections) {
...
-            return existingConnection;
-          }
-        });
-
-    if (!isNull(disconnectAction.get())) {
-      disconnectAction.get().run();
-    }
-    if (newConnectionAccepted.get()) {
+    if (checkWhetherToConnect(
+        peer, true)) { // TODO: checkWhetherToConnect should return Optional<DisconnectReason>
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java:347]: will this TODO be part of this PR or separate?

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java [hunk 8]
```diff
@@ -1098,4 +601,16 @@ public void testDispatchUpdate(
       this.dispatchUpdate(permissionsRestricted, affectedPeers);
     }
   }
+
+  private static class ConnectionDispatch {
+    private int dispatchNo;
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java:606]:
```suggestion
    private int dispatchCount;
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java [hunk 9]
```diff
@@ -1098,4 +601,16 @@ public void testDispatchUpdate(
...
+
+    public void onConnect(final PeerConnection pc) {
+      this.dispatchNo++;
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java:609]:
```suggestion
      this.dispatchCount++;
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java [hunk 10]
```diff
@@ -1098,4 +601,16 @@ public void testDispatchUpdate(
...
+    }
+
+    public int getDispatchNo() {
+      return dispatchNo;
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java:613]:
```suggestion
      return dispatchCount;
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java [hunk 11]
```diff
@@ -1098,4 +601,16 @@ public void testDispatchUpdate(
...
   }
+
+  private static class ConnectionDispatch {
+    private int dispatchNo;
+
+    public void onConnect(final PeerConnection pc) {
+      this.dispatchNo++;
+    }
+
+    public int getDispatchNo() {
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java:612]:
```suggestion
    public int getDispatchCount() {
```

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java [hunk 12]
```diff
@@ -170,16 +173,17 @@ public void connect_toDiscoveryPeerUpdatesStats() {
...
+
     final Peer peer = createPeer();
     final PeerConnection connection = connection(peer);
     connectionInitializer.simulateIncomingConnection(connection);
 
-    final Optional<CompletableFuture<PeerConnection>> trackedConnection =
-        agent.getPeerConnection(peer);
-    assertThat(trackedConnection).isNotEmpty();
-    assertThat(trackedConnection.get()).isCompletedWithValue(connection);
+    assertThat(connectionDispatch.getDispatchNo()).isEqualTo(1);
```
Comment (macfarla) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java:186]:
```suggestion
    assertThat(connectionDispatch.getDispatchCount()).isEqualTo(1);
```

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 13]
```diff
@@ -102,48 +125,98 @@ public EthPeers(
...
+    final EthPeer peer = connections.get(id);
+    return registerDisconnect(id, peer, connection);
+  }
+
+  private boolean registerDisconnect(
+      final Bytes id, final EthPeer peer, final PeerConnection connection) {
+    notReadyConnections.invalidate(connection);
+    boolean removed = false;
+    if (peer != null && peer.getConnection().equals(connection)) {
+      if (!peerHasNonReadyConnection(id)) {
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:182]: is there a better name than "non ready" - incomplete? partial?

File: besu/src/test/java/org/hyperledger/besu/RunnerTest.java [hunk 14]
```diff
@@ -260,7 +261,7 @@ private void syncFromGenesis(final SyncMode mode, final GenesisConfigFile genesi
       final OkHttpClient client = new OkHttpClient();
       Awaitility.await()
           .ignoreExceptions()
-          .atMost(2L, TimeUnit.MINUTES)
+          .atMost(5L, TimeUnit.MINUTES)
```
[the remark this answers is not in the document]
Comment (pinges) [reply besu/src/test/java/org/hyperledger/besu/RunnerTest.java:264]: reverted to 2!

File: consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/protocol/BftProtocolManager.java [hunk 15]
```diff
@@ -107,6 +108,11 @@ public void handleNewConnection(final PeerConnection peerConnection) {
     peers.add(peerConnection);
   }
 
+  @Override
+  public boolean shouldConnect(final Peer peer, final boolean incoming) {
+    return false; // for now the EthProtocolManager takes care of this
```
[the remark this answers is not in the document]
Comment (pinges) [reply consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/protocol/BftProtocolManager.java:113]: Yes! Now that we can decide to connect on this layer we should make use of it!

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java [hunk 16]
```diff
@@ -228,22 +232,37 @@ public RequestManager.ResponseStream send(final MessageData messageData) throws
...
+   * multiple connections to the same peer at that time
+   *
+   * @param messageData the data to send
+   * @param protocolName the protocol to use for sending
+   * @param conToUse the connection to use for sending
+   * @return the response stream from the peer
+   * @throws PeerNotConnected if the peer is not connected
+   */
+  public RequestManager.ResponseStream send(
+      final MessageData messageData, final String protocolName, final PeerConnection conToUse)
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:249]: changed it to connectionToUse

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java [hunk 17]
```diff
@@ -228,22 +232,37 @@ public RequestManager.ResponseStream send(final MessageData messageData) throws
...
+    return send(messageData, protocolName, this.connection);
+  }
+
+  /**
+   * This method is only used for sending the status message, as it is possible that we have
+   * multiple connections to the same peer at that time
+   *
+   * @param messageData the data to send
+   * @param protocolName the protocol to use for sending
+   * @param conToUse the connection to use for sending
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:244]: This is in line 239-240 :-)

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java [hunk 18]
```diff
@@ -590,6 +625,41 @@ public Optional<BlockHeader> getCheckpointHeader() {
...
+   * @param a The first connection
+   * @param b The second connection
+   * @return A negative value if {@code a} should be kept, a positive value is {@code b} should be
+   *     kept
+   */
+  private int compareDuplicateConnections(final PeerConnection a, final PeerConnection b) {
+
+    if (a.isDisconnected() != b.isDisconnected()) {
+      // One connection has failed - prioritize the one that hasn't failed
+      return a.isDisconnected() ? 1 : -1;
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:644]: This method is called right after we have received the status message, so we should be connected. But even if the connection is disconnected in the meantime the connections will be cleaned up ... This method just selects the one that should be used of the two possible ones.

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java (hunk 9 above)
Comment (pinges) [reply ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgentTest.java:609]: Done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java (hunk 13 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:182]: chaged it to incomplete! Far better!

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 19]
```diff
@@ -58,40 +67,54 @@ public class EthPeers {
...
-  private final int maxPeers;
   private final int maxMessageSize;
   private final Subscribers<ConnectCallback> connectCallbacks = Subscribers.create();
   private final Subscribers<DisconnectCallback> disconnectCallbacks = Subscribers.create();
   private final Collection<PendingPeerRequest> pendingRequests = new CopyOnWriteArrayList<>();
+  private final int peerLowerBound;
+  private final int peerUpperBound;
+  private final int maxRemotelyInitiatedConnections;
+  private final Boolean randomPeerPriority;
+  private final Bytes nodeIdMask = Bytes.random(64);
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:89]: in the original this was a constant - message length? can we avoid hard coding it

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 20]
```diff
@@ -250,6 +324,47 @@ public Comparator<EthPeer> getBestChainComparator() {
...
+
+  private Stream<PeerConnection> getAllActiveConnections() {
+    return connections.values().stream()
+        .map(EthPeer::getConnection)
+        .filter(c -> !c.isDisconnected());
+  }
+
+  private Stream<PeerConnection> getAllConnections() {
+    return Stream.concat(
+            connections.values().stream().map(EthPeer::getConnection),
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:343]: completeConnections ?

Comment (macfarla) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:343]: to complement incompleteConnections

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/wire/ShouldConnectCallback.java [hunk 21]
```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright ConsenSys AG.
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/wire/ShouldConnectCallback.java:2]: Done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java (hunk 19 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:89]: Done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java (hunk 20 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:343]: Done

**Review by fab-10 — COMMENTED** [review]
overall looks, I have some comment and suggestions, and if possible I would like to know some of the future developments/features that could be built on top of this one.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java [hunk 22]
```diff
@@ -590,6 +630,41 @@ public Optional<BlockHeader> getCheckpointHeader() {
...
+    }
+
+    final Bytes peerId = a.getPeer().getId();
+    // peerId is the id of the other node
+    if (a.inboundInitiated() != b.inboundInitiated()) {
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:654]: Here we have to make sure that both sides keep the same connection. If each side have initiated one of the connections we keep the one initiated by the peer with the lower id (public key).

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 23]
```diff
@@ -115,48 +136,97 @@ public EthPeers(
...
+                    newConnection,
+                    protocolName,
+                    this::ethPeerStatusExchanged,
+                    peerValidators,
+                    maxMessageSize,
+                    clock,
+                    permissioningProviders,
+                    localNodeId));
+      }
+      incompleteConnections.put(newConnection, ethPeer);
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:171]: We do have an incomplete connection with this peer. The problem is that both sides have initiated one at about the same time. We do check whether we have a completed connection before we try to create a new connection (_shouldConnect()_ in EthPeers). Each side might complete a different connection first. To make sure that the connection succeeds we give both connections a chance to complete, and if both of them do complete, we choose the one we keep based on the method _compareDuplicateConnections()_ in EthPeers.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 24]
```diff
@@ -294,18 +408,174 @@ public interface ConnectCallback {
...
+    return a.getPeer().getId().xor(nodeIdMask).compareTo(b.getPeer().getId().xor(nodeIdMask));
+  }
+
+  private void enforceRemoteConnectionLimits() {
+    if (!shouldLimitRemoteConnections() || peerCount() < maxRemotelyInitiatedConnections) {
+      // Nothing to do
+      return;
+    }
+
+    getActivePrioritizedPeers()
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:466]: You are right. This is doing what it did before my PR and it should be changed. We should definitely change that to take the reputation into account.
I think we should do that in a separate PR.

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/peers/DefaultPeer.java [hunk 25]
```diff
@@ -60,6 +63,16 @@ public EnodeURL getEnodeURL() {
...
   }
 
+  @Override
+  public Optional<ForkId> getForkId() {
+    return Optional.ofNullable(forkId);
+  }
+
+  @Override
+  public void setForkId(final ForkId forkId) {
+    this.forkId = forkId;
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/peers/DefaultPeer.java:73]: There are two places where we can get the fork id. If you enable --Xfilter-on-enr-fork-id=true we will ask for the ENR during discovery. The ENR might contain the fork id. The other place is when we receive the status message (see EthProtocolManager). It is not know when we create the object, unfortunately.

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/peers/Peer.java [hunk 26]
```diff
@@ -47,4 +51,29 @@ static Bytes randomId() {
...
+  default Optional<ForkId> getForkId() {
+    return Optional.empty();
+  }
+
+  /**
+   * Sets the fork id
+   *
+   * @param forkId The fork id
+   */
+  void setForkId(ForkId forkId);
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/peers/Peer.java:78]: s.a.

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java [hunk 27]
```diff
@@ -348,36 +303,29 @@ private CompletableFuture<PeerConnection> initiateOutboundConnection(final Peer
...
+    return peerPrivileges.canExceedConnectionLimits(peer);
+  }
+
+  public void setGetAllConnectionsCallback(final Callable<Stream<PeerConnection>> callback) {
+    this.getAllConnectionsCallback = callback;
+  }
+
+  public void setGetAllActiveConnectionsCallback(final Callable<Stream<PeerConnection>> callback) {
+    this.getAllActiveConnectionsCallback = callback;
+  }
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java:316]: You are right, Supplier is much nicer! Changed that! Unfortunately the RlpxAgent is set up in the Runner and the EthPeers is set up in the Controller which means that the suppliers are not available when the class is created.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 28]
```diff
@@ -115,48 +136,97 @@ public EthPeers(
...
+    final EthPeer peer = completeConnections.get(id);
+    return registerDisconnect(id, peer, connection);
+  }
+
+  private boolean registerDisconnect(
+      final Bytes id, final EthPeer peer, final PeerConnection connection) {
+    incompleteConnections.invalidate(connection);
+    boolean removed = false;
+    if (peer != null && peer.getConnection().equals(connection)) {
+      if (!peerHasIncompleteConnection(id)) {
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:193]: The connection we are disconnecting is not necessarily the connection that is associated with the EthPeer. That is what we are checking in line 192. In line 190 we are removing the connection from the incompleteConnections, if it is in that cache.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java [hunk 29]
```diff
@@ -98,9 +98,10 @@ protected boolean removeEldestEntry(final Map.Entry<Hash, Boolean> eldest) {
 
   private final Map<String, Map<Integer, RequestManager>> requestManagers;
 
-  private final AtomicReference<Consumer<EthPeer>> onStatusesExchanged = new AtomicReference<>();
+  private final Consumer<EthPeer> onStatusesExchanged;
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:101]: Reintroducing the old logic that prevented the callback to be called only once does help, I guess. I will reintroduce it!

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 30]
```diff
@@ -294,18 +408,174 @@ public interface ConnectCallback {
...
+        .filter(p -> !p.isDisconnected())
+        .sorted(this::comparePeerPriorities);
+  }
+
+  private void enforceConnectionLimits() {
+    if (peerCount() < peerUpperBound) {
+      // Nothing to do - we're under our limits
+      return;
+    }
+    getActivePrioritizedPeers()
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:491]: s.a.

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/peers/DefaultPeer.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (fab-10) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/peers/DefaultPeer.java:73]: do you think we can enable `--Xfilter-on-enr-fork-id` by default? Probably in the next quarter release 23.4?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (fab-10) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:656]: Do you know which strategy other clients are using? I mean could the other client close the connection that we are keeping open?

File: besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java [hunk 31]
```diff
@@ -684,13 +634,13 @@ public Runner build() {
...
-            .setClientId(BesuInfo.nodeName(identityString))
-            .setLimitRemoteWireConnectionsEnabled(limitRemoteWireConnectionsEnabled)
-            .setFractionRemoteWireConnectionsAllowed(fractionRemoteConnectionsAllowed);
-    networkingConfiguration.setRlpx(rlpxConfiguration).setDiscovery(discoveryConfiguration);
+            .setClientId(BesuInfo.nodeName(identityString));
+    networkingConfiguration
+        .setRlpx(rlpxConfiguration)
+        .setDiscovery(
+            discoveryConfiguration); // TODO: Check whether this is OK, as we do getDiscovery in
+    // line 622 and 624 above ...
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java:643]: what are the implications of this TODO

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 32]
```diff
@@ -1320,6 +1323,13 @@ static class TxPoolOptionGroup {
           "Specifies the maximum number of blocks to retrieve logs from via RPC. Must be >=0. 0 specifies no limit  (default: ${DEFAULT-VALUE})")
   private final Long rpcMaxLogsRange = 5000L;
 
+  @CommandLine.Option(
+      hidden = true,
+      names = {"--p2p-peer-lower-bound"}, // TODO: do we want to make this not experimental
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1328]: if so - needs a changelog - this is a breaking change - and docs

File: besu/src/main/resources/log4j2.xml [hunk 33]
```diff
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <Configuration status="WARN">
     <Properties>
-        <Property name="root.log.level">${env:LOG_LEVEL:-INFO}</Property>
+        <Property name="root.log.level">${env:LOG_LEVEL:-DEBUG}</Property>
+        <!--Revert this to INFO before merging into main-->
```
Comment (macfarla) [inline besu/src/main/resources/log4j2.xml:5]: TODO!

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (macfarla) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeer.java:655]: this is spec'd somewhere, right @pinges ?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/TrailingPeerLimiter.java [hunk 34]
```diff
@@ -65,7 +65,7 @@ public void enforceTrailingPeerLimit() {
     while (!trailingPeers.isEmpty() && trailingPeers.size() > maxTrailingPeers) {
       final EthPeer peerToDisconnect = trailingPeers.remove(0);
       LOG.debug("Enforcing trailing peers limit by disconnecting {}", peerToDisconnect);
-      peerToDisconnect.disconnect(DisconnectReason.TOO_MANY_PEERS);
+      peerToDisconnect.disconnect(DisconnectReason.USELESS_PEER);
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/TrailingPeerLimiter.java:68]: why this change?

File: .circleci/config.yml [hunk 35]
```diff
@@ -401,6 +401,9 @@ workflows:
       - acceptanceTests:
           requires:
             - assemble
+      - acceptanceTestsNonMainnet:
+          requires:
+            - assemble
```
Comment (macfarla) [inline .circleci/config.yml:406]: TODO :)

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/NettyConnectionInitializer.java [hunk 36]
```diff
@@ -220,6 +220,15 @@ private ChannelInitializer<SocketChannel> inboundChannelInitializer() {
...
       protected void initChannel(final SocketChannel ch) throws Exception {
+        // TDOD: check whether we want to connect to this peer. If we can find this peer in the
+        // PeerTable call the "checkWhetherToConnect" method in RlpxAgent
+        // we do have access to the ip address here (from the channel) and, if forkId is enabled, we
+        // have a good chance that the peers in the PeerTable contain the IP address (as part of the
+        // NodeRecord), so we can match them
+        // If the IP address is available we have a chance that the ForkId is available for that
+        // peer as well, which means that know whether the peer is on the right chain.
+        // Doing that decission here in the inboundChannelnitializer means that we could sort out
+        // peers on a wrong chain without having to do the handshake.
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/NettyConnectionInitializer.java:231]: can we create an issue for this to replace the TODO

**Review by macfarla — COMMENTED** [review]
looking good. few comments

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java (hunk 27 above)
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/RlpxAgent.java]: Done

File: besu/src/main/resources/log4j2.xml (hunk 33 above)
Comment (pinges) [reply besu/src/main/resources/log4j2.xml:5]: Done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/TrailingPeerLimiter.java (hunk 34 above)
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/TrailingPeerLimiter.java:68]: I think this is the right thing to do!

File: .circleci/config.yml (hunk 35 above)
Comment (pinges) [reply .circleci/config.yml:406]: Done

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/NettyConnectionInitializer.java (hunk 36 above)
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/rlpx/connections/netty/NettyConnectionInitializer.java:231]: Done

**Review by fab-10 — COMMENTED** [review]
it is easy to get lost in the peering code! Based on my previous review and on the fact that you have tested this for a long time, for me is ok to go on and approve, I just see that there are still some open questions from @macfarla.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 32 above)
Comment (pinges) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: DONE

**Review by macfarla — APPROVED** [review]
LGTM

**Comment by macfarla** [comment]
Would be good to get a second approval @fab-10  :)
